### ISSUE TITLE
Fix an API definition

### ISSUE THREAD

**Issue by whong92** [whong92/synth-net-render#1]
Discussion for API definition, and general structure of project. Aims:

- [ ] Define top-level API for synth-net-render (top-level class, methods, arguments etc), so that it satisfies the following:
    - backend agnostic (does not depend on rendering engine, or underlying implementation)
    - cover as many use cases as possible
    - prevent/limit future changes to API as much as possible

- [ ] Figure out way to structure project to make it extensible. This involves separating out the blender code to general purpose stuff like RandomLib and SceneLib etc. Think about use cases such as adding a new graphics engine (Unrel/WebGL etc), or parallelizing rendering jobs for example.

**Comment by whong92** [comment]
I am going to kick this off by proposing a few use cases, which can perhaps help in defining what the API should look like:
 - Simplest : single-subject rendering with random background image, no extra props. (What we already have)
 - Intermediate : single-subject rendering with extra props (Flying distractors, virtual rendered background, kind of like what was done here : https://arxiv.org/pdf/1804.06516.pdf). Offers the option of having a purely virtual background.
 - Expert : multi-subject rendering. Self explanatory. Can be quite interesting, depending on how users want to customize scenes.
- All of the use cases above supports what we already have (randomization of different hyperparameters, with the option of making things more complex)

I think these different use cases can be quite nicely encapsulated using a `renderSession` object. Where a session is an object provided with a list of subjects, and rendering hyperparameters:
```
renderSession(['obj1.model', 'obj2.model', 'obj3.model'], hyperparameter_dict)
```
For each `renderSession`, the images will be produced using the hyperparameters, with some subset of the list of the models provided (multi-subject rendering).

**Comment by whong92** [comment]
In addition, `renderSession` should be able to provide methods that suggest the possibility of different workflows of generating images (e.g. generating a single image on-demand, generating all the images, generating batches of images). This emulates the `ImageGenerator` API that is provided in the Keras API: `https://keras.io/preprocessing/image/`

In some implementations, this might be difficult or impossible, (for example in our current implementation, doing this would involve re-starting the blender subprocess every time we need a new image), in which case we throw a `NotSupportedError` error or something like that.

**Comment by 921kiyo** [comment]
How about we document the simplest API and clean up the code accordingly (this should be minimal work to begin with)? Later, everytime we add extra stuff, we document first and change the code iteratively. I feel it is nice to have a very small prototype to get this going:)

**Comment by whong92** [comment]
@921kiyo I agree the implementation should be the simplest (we start with what we already have). However, the user-facing API needs to accomodate as many use cases as possible, and be as general as possible (though it need not have the actual implementation of all use cases - thus we can throw a `NotSupportedError` when some unsupported feature is called). 

Doing this will help us avoid re-writing the whole code base everytime we make a small change / add a new feature. Besides that, from experience, it can help guide the design, encourage re-use of code, and constantly remind us what we need to deliver.

**Comment by Kroupavel** [comment]
You mention the API being agnostic of the underlying rendering engine. So far we have only one engine, right? What are the others you were thinking of? Or by different rendering engine you mean the different rendering settings (interior, exterior, lightning normalisation)?
